<a href="https://colab.research.google.com/github/hemareddyyanala/HemaReddy_INFO5731_Fall2024/blob/main/Yanala_Hema_Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [ ]:
# Your code here
import math

# Formula for the volume of a hemisphere: V = (2/3) * π * r^3

def calculate_volume_of_hemisphere(radius):
    return (2/3) * math.pi * radius**3

# Print header
print("r    V")

# Loop through even numbers from 1 to 20
for r in range(2, 21, 2):
    V = calculate_volume_of_hemisphere(r)
    print(f"{r:<4} {V:.2f}")




r    V
2    16.76
4    134.04
6    452.39
8    1072.33
10   2094.40
12   3619.11
14   5747.02
16   8578.64
18   12214.51
20   16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [ ]:
# Write code for each of the sub parts with proper comments.
import random

# Generate and sort 12 random numbers in the range of 50 to 150
numbers = [random.randint(50, 150) for _ in range(12)]
asc_numbers = sorted(numbers)
desc_numbers = sorted(numbers, reverse=True)

# Calculate averages
avg_all = sum(numbers) / len(numbers)
avg_first_6_asc = sum(asc_numbers[:6]) / 6
avg_first_6_desc = sum(desc_numbers[:6]) / 6

# Display results
print("Generated Numbers:", numbers)
print("Ascending Order:", asc_numbers)
print("Descending Order:", desc_numbers)
print(f"Average of All Numbers: {avg_all:.2f}")
print(f"Average of First 6 (Ascending): {avg_first_6_asc:.2f}")
print(f"Average of First 6 (Descending): {avg_first_6_desc:.2f}")


Generated Numbers: [122, 123, 134, 106, 84, 89, 84, 86, 130, 87, 56, 68]
Ascending Order: [56, 68, 84, 84, 86, 87, 89, 106, 122, 123, 130, 134]
Descending Order: [134, 130, 123, 122, 106, 89, 87, 86, 84, 84, 68, 56]
Average of All Numbers: 97.42
Average of First 6 (Ascending): 77.50
Average of First 6 (Descending): 117.33


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [ ]:
# Your code here

# (a) Prompt the user to input two sentences
sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

# (b) Split sentences into lists of words
words1 = sentence1.split()
words2 = sentence2.split()
print("Word List:", words1 + words2)

# (c) Find common words between the two sentences
common_words = list(set(words1) & set(words2))
print("Common Words:", common_words)

# (d) Remove punctuation and print sanitized sentences
punctuation = ".,;:!?"
sanitized1 = ''.join(char for char in sentence1 if char not in punctuation)
sanitized2 = ''.join(char for char in sentence2 if char not in punctuation)
print("Sanitized Sentence 1:", sanitized1)
print("Sanitized Sentence 2:", sanitized2)

# (e) Count the number of vowels in each sentence
vowels = "aeiou"
count_vowels = lambda s: {v: s.lower().count(v) for v in vowels}
vowel_count1 = count_vowels(sanitized1)
vowel_count2 = count_vowels(sanitized2)
print("Vowels in Sentence 1:", vowel_count1)
print("Vowels in Sentence 2:", vowel_count2)


Enter the first sentence: Are you okay?
Enter the second sentence: Are you happy?
Word List: ['Are', 'you', 'okay?', 'Are', 'you', 'happy?']
Common Words: ['you', 'Are']
Sanitized Sentence 1: Are you okay
Sanitized Sentence 2: Are you happy
Vowels in Sentence 1: {'a': 2, 'e': 1, 'i': 0, 'o': 2, 'u': 1}
Vowels in Sentence 2: {'a': 2, 'e': 1, 'i': 0, 'o': 1, 'u': 1}


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [ ]:
# Your code here

import math

def calculate(expression):
    try:
        result = eval(expression, {"__builtins__": None}, {
            "sqrt": math.sqrt,
            "pow": math.pow,
            "mod": lambda x, y: x % y
        })
        return result
    except ZeroDivisionError:
        return "Error: Division by zero!"
    except Exception:
        return "Error: Invalid input!"

def main():
    while True:
        expr = input("Enter expression (or 'exit' to quit): ")
        if expr.lower() == 'exit':
            break
        print("Result:", calculate(expr))

if __name__ == "__main__":
    main()


Enter expression (or 'exit' to quit): 6**2
Result: 36
Enter expression (or 'exit' to quit): sqrt(144)
Result: 12.0


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [4]:
# Your code here

import pandas as pd
import numpy as np

# Load data into a DataFrame
df = pd.read_csv('/content/Assignment1_denton_housing.csv')

# Pivot the DataFrame for easier access
pivot_df = df.pivot(index='year', columns='title_field', values='value')

# (a) Calculate year-to-year differences in Occupied Housing Units
occupied_diff = pivot_df['Occupied Housing Units'].diff().dropna().sort_values()
print("Year-to-Year Differences in Occupied Housing Units:", occupied_diff.tolist())

# (b) Calculate percentages of Vacant and Occupied Housing Units
total_units = pivot_df.sum(axis=1)
percentages = pivot_df.div(total_units, axis=0) * 100
print("\nYearly Percentages:\n", percentages[['Vacant Housing Units', 'Occupied Housing Units']].round(2))

# (c) Find years with the highest Vacant and Occupied Housing Units
max_vacant_year = pivot_df['Vacant Housing Units'].idxmax()
max_occupied_year = pivot_df['Occupied Housing Units'].idxmax()
print("\nMax Vacant Year:", max_vacant_year, pivot_df.loc[max_vacant_year, 'Vacant Housing Units'])
print("Max Occupied Year:", max_occupied_year, pivot_df.loc[max_occupied_year, 'Occupied Housing Units'])

# (d) Calculate harmonic mean of total housing units
harmonic_mean_total_units = len(total_units) / np.sum(1 / total_units)
print("\nHarmonic Mean of Total Housing Units:", round(harmonic_mean_total_units, 2))



Year-to-Year Differences in Occupied Housing Units: [-795.0, -303.0, 1404.0, 1565.0, 2622.0, 5091.0]

Yearly Percentages:
 title_field  Vacant Housing Units  Occupied Housing Units
year                                                     
2008                         3.74                   30.63
2009                         3.10                   31.05
2010                         2.77                   31.29
2011                         3.21                   30.99
2012                         2.26                   31.61
2013                         3.00                   31.12
2014                         1.94                   31.84

Max Vacant Year: 2008 4483
Max Occupied Year: 2014 46295

Harmonic Mean of Total Housing Units: 130257.98


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# Write your response below
'''
I think the most challenging part was to keep updating and changing the program to solve every sub-question.
I enjoyed using pandas to solve the last question. I think the time provided to submit the assignment is sufficient.
I think the assignment was great in terms of the practical exposure I got. I did a good amount of research about the
different python functions I could to solve different questions and used stackoverflow when I got errors. This is a
great start for the programming journey.
'''